### PR TITLE
[Merged by Bors] - feat: add algebra of sets

### PR DESCRIPTION
**Pull request by EtienneC30**
Define the notion of algebra of sets and prove basic properties.
Define the algebra of set generated by a family and prove basic properties.
Prove that the algebra of sets generated by a countable family is countable.

---

This is the first PR involving a whole new file that I do on my own, with new definitions and the use of namespaces. I tried to inspire from the similar files `PiSystem` and `SetSemiring`.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/MeasureTheory/SetAlgebra.lean [hunk 1]
```diff
@@ -0,0 +1,227 @@
+/-
+Copyright (c) 2024 Etienne Marion. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Etienne Marion
+-/
+import Mathlib.MeasureTheory.SetSemiring
```
Comment (eric-wieser) [inline Mathlib/MeasureTheory/SetAlgebra.lean:6]: Is there a link between `IsSetSemiring` and `IsSetAlgebra` that is easy to formalize?

File: Mathlib/MeasureTheory/SetAlgebra.lean [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+by the algebra of sets generated by `𝒜`. -/
+theorem generateFrom_generateSetAlgebra_eq :
+    generateFrom 𝒜 = generateFrom (generateSetAlgebra 𝒜) := by
+  refine le_antisymm (generateFrom_mono self_subset_generateSetAlgebra) (fun s ms ↦ ?_)
+  refine @generateFrom_induction _ _ (generateSetAlgebra 𝒜) (fun t ht ↦ ?_)
+    (@MeasurableSet.empty _ (generateFrom 𝒜))
+    (fun t ↦ MeasurableSet.compl)
+    (fun f hf ↦ MeasurableSet.iUnion hf)
+    s ms
+  induction' ht with u u_mem u _ mu u v _ _ mu mv
```
Comment (eric-wieser) [inline Mathlib/MeasureTheory/SetAlgebra.lean:125]: Could you switch this to the unprimed `induction` tactic, which is a bit more readable?

**Review by sgouezel — COMMENTED** [review]
Looks good!

File: Mathlib/MeasureTheory/SetAlgebra.lean [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+  | compl u _ mu => exact mu.compl
+  | union u v _ _ mu mv  => exact MeasurableSet.union mu mv
+
+/-- If a family of sets `𝒜` is contained in `ℬ`, then the algebra of sets generated by `𝒜`
+is contained in the one generated by `ℬ`. -/
+theorem generateSetAlgebra_mono {ℬ : Set (Set α)} (h : 𝒜 ⊆ ℬ) :
+    generateSetAlgebra 𝒜 ⊆ generateSetAlgebra ℬ := by
+  intro s hs
+  induction hs with
+  | @base t t_mem => exact self_subset_generateSetAlgebra (h t_mem)
```
Comment (sgouezel) [inline Mathlib/MeasureTheory/SetAlgebra.lean:142]: You're still using the `@` version here and below (several times). Is that on purpose?

Comment (EtienneC30) [reply Mathlib/MeasureTheory/SetAlgebra.lean]: Sorry I forgot them

**Comment by sgouezel** [comment]
bors r+
Thanks!

**Pull request closed without merging**